### PR TITLE
add github workflow for CI and website,

### LINKED ISSUE #1847: Leverage Github Actions
## 🚀 Feature Proposal

We have Github Actions enabled in this repo! What can we do with them?
One thing that would be nice is to move the trigger to CircleCI for deploying the website.
Moreover, we can actually redo the automation for the [website](https://github.com/fastify/website) using actions!

@lmammino and maybe others, would it be nice to have these here?

### PR DESCRIPTION
**Pull request by StarpTech**
Fix #1847 This PR will add basic support for github workflows. There are some questions:

- [ ] Do we really need to test YARN?
- [ ] Why do we use a java test coverage tool "cobertura" ?
- [ ] I would like to remove all other CI tools in that PR. Any objections?

@mcollina you have to add some secrets. You can find it in the settings of this repo.

- `circleci_token` to trigger the website build

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [X] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      - name: Check licenses
+        run: |
+          npm run license-checker
+      - name: Run tests
+        run: |
+          npm run test:report
+      - name: Generate coverage
+        run: |
+          npm run coveralls -- -c
+      - name: Coveralls
```
Comment (jsumners) [inline .github/workflows/ci.yml]: I'm literally working on doing this in the ldapjs repo right now. You need to use the parallel version -- https://github.com/coverallsapp/github-action#complete-parallel-job-example

Comment (StarpTech) [reply .github/workflows/ci.yml]: Only for performance or what's the issue?

Comment (StarpTech) [reply .github/workflows/ci.yml]: I got it due to matrix.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [8.x, 10.x, 12.x]
```
Comment (Eomm) [inline .github/workflows/ci.yml]:
```suggestion
        node-version: [6.x, 8.x, 10.x, 12.x]
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        os: [ubuntu-latest, windows-latest, macOS-latest]
+
+    steps:
+      - uses: actions/checkout@v1
+      - name: Use Node.js ${{ matrix.node-version }}
```
Comment (Eomm) [inline .github/workflows/ci.yml]:
```suggestion
      - name: Use Node.js ${{ matrix.node-version }} in ${{ matrix.os }}
```

File: .github/workflows/website.yml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+name: website
+
+on:
+  pull_request:
+    branches:
+      - master
```
Comment (Eomm) [inline .github/workflows/website.yml]: As before: updating the site every pull_request event that changes docs/ isn't too much?
the push event seems enough for me

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+name: ci
+
+on: [push, pull_request]
```
Comment (Eomm) [inline .github/workflows/ci.yml]: I'm not sure pull_request fit this workflow.
Reading [this doc](https://help.github.com/en/articles/events-that-trigger-workflows) I think we are updating the coverage badge too often right?

What do you think about to split this in two workflows, one dedicated for push to master that update the coverage and one push that trigger tests? I say push because of [this suggestion](https://help.github.com/en/articles/events-that-trigger-workflows#pull-request-events-for-forked-repositories)

Comment (jsumners) [reply .github/workflows/ci.yml]: I really don't understand what that note is saying. Is it saying that a PR from a forked repository will not run tests on the forked code when the `pull_request` event is received?

Comment (jsumners) [reply .github/workflows/ci.yml]: That does not seem to be the case -- https://github.com/ldapjs/node-ldapjs/pull/553/checks?check_run_id=222806190#step:5:466

The test that link is referencing only exists in the forked repo.

Comment (Eomm) [reply .github/workflows/ci.yml]: > The test that link is referencing only exists in the forked repo.

Thanks, I didn't get that, so the `pull_request` is needed, gotcha 👍

My concern is about this use case: if we ask for a review or add a label the `pull_request` event is triggered and the step `Generate coverage` is executed.
I think the update of the badges could happen only on push on master

Comment (jsumners) [reply .github/workflows/ci.yml]: This might solve it:

```suggestion
on:
  - push
  - pull_request:
    types:
    - synchronized
```

Comment (StarpTech) [reply .github/workflows/ci.yml]: @jsumners what does `types` means?

Comment (StarpTech) [reply .github/workflows/ci.yml]: Nevermind, nice tip,

Comment (StarpTech) [reply .github/workflows/ci.yml]: @jsumners @Eomm check here https://help.github.com/en/articles/events-that-trigger-workflows the `pull_request` is only triggered opened, synchronize, reopened that should be fine.

Comment (jsumners) [reply .github/workflows/ci.yml]: I understand that table now. Man, whomever is writing the documentation for Actions has a very weird and hard to follow writing style.

Yeah, just stick with `pull_request`. No need to declare the `types` that trigger it.

**Comment by jsumners** [comment]
I haven't had time to work up a reproducible test case for the `fastify-plugin` + `pnpm` misleading warning.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (jsumners) [reply .github/workflows/ci.yml]: Oh, I missed this. You should definitely be supplying the matrix values to this `runs-on`:

https://github.com/ldapjs/node-ldapjs/blob/a113953e0d91211eb945d2a3952c84b7af6de41c/.github/workflows/main.yml#L22-L32

File: .github/workflows/ci.yml (hunk 5 above)
Comment (Eomm) [reply .github/workflows/ci.yml]: Clear, thanks for the check

File: .github/workflows/ci.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply .github/workflows/ci.yml]: @mcollina https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idruns-on

File: .github/workflows/package-manager-ci.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
+name: package-manager-ci
+
+on:
+  push:
+    branches:
+      - master
+
+jobs:
+  pnpm:
+    runs-on: ubuntu-latest
```
Comment (jsumners) [inline .github/workflows/package-manager-ci.yml]:
```suggestion
    runs-on: ${{ matrix.os }}
```

Otherwise you'll only be running these on `ubuntu-latest` instead of the matrix specified `ubuntu-16.04`.

**Comment by StarpTech** [comment]
Github actions can't be enabled with a PR so some bugs must be fixed afterwards.

**Comment by StarpTech** [comment]
I found out that Container Actions are only supported on Linux :expressionless: so we can't reuse them on MAC or WIN. We would need to write Javascript actions or run the scripts directly on the virtual environment.

**Comment by StarpTech** [comment]
In our case we use only the `coveralls` action in cross OS environemnts and it's a javascript action. Lucky shoot!

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    runs-on: ${{ matrix.os }}
+
+    strategy:
+      matrix:
+        node-version: [6.x, 8.x, 10.x, 12.x]
+        os: [ubuntu-16.04, windows-2016, macOS-10.14]
+
+    steps:
+      - uses: actions/checkout@v1
+      - name: Use Node.js ${{ matrix.node-version }} in ${{ matrix.os }}
```
Comment (jsumners) [inline .github/workflows/ci.yml]: BTW, the system automatically appends the Node and OS version. So if this were `name: foo` then the tests would show up like `foo (ubuntu-16.04, node-10.x)`.

See https://github.com/ldapjs/node-ldapjs/runs/222797629#step:1:1

**Comment by StarpTech** [comment]
@delvedor this adds noise. We can enable them at any time in seconds @mcollina won't delete them. Let's give Github Actions a try and we will see.

**Comment by delvedor** [comment]
@StarpTech I'm fine with the noise.
As said GitHub Action is in beta, and have the two systems in parallel for a while (1/2 months max) will help us make sure that we can trust GitHub Actions.
First of all, we should guarantee that our test are passing and the software is stable, then, our maintainers experience.

**Comment by mcollina** [comment]
I've added the `CIRCLECI_TOKEN` secret.
